src/auth/jsonfile.rs

```diff
@@ -0,0 +1,69 @@
+use crate::auth::*;
+
+use futures::Future;
+
+use std::fs;
+use serde::Deserialize;
+use serde_json;
+
+use tokio::timer::Delay;
+use std::time::{Duration, Instant};
+
+#[derive(Deserialize, Clone, Debug)]
+struct Credentials {
+    username: String,
+    password: String,
+}
+
+/// [`Authenticator`] implementation that authenticates against a JSON file.
+///
+/// [`Authenticator`]: ../trait.Authenticator.html
+#[derive(Clone, Debug)]
+pub struct JsonfileAuthenticator {
+    credentials_list: Vec<Credentials>,
+}
+
+impl JsonfileAuthenticator {
+    /// Initialize a new [`JsonfileAuthenticator`] from file.
+    pub fn new(filename: String) -> Self {
+        let s = fs::read_to_string(filename).unwrap();
+        let credentials_list: Vec<Credentials> = serde_json::from_str(&s).unwrap();
+        JsonfileAuthenticator { credentials_list }
+    }
+}
+
+impl Authenticator<AnonymousUser> for JsonfileAuthenticator {
+    fn authenticate(&self, _username: &str, _password: &str) -> Box<dyn Future<Item = AnonymousUser, Error = ()> + Send> {
+        let username = _username.to_string();
+        let password = _password.to_string();
+        let credentials_list = self.credentials_list.clone();
+
+        Box::new(
+            futures::future::ok(())
+                .and_then(move |_| {
+                    for c in credentials_list.iter() {
+                        if username == c.username {
+                            if password == c.password {
+                                info!("Successful login by user {}", username);
+                                return Ok(AnonymousUser {});
+                            } else {
+                                warn!("Failed login for user {}: bad password", username);
+                                return Err(());
+                            }
+                        }
+                    }
+                    warn!("Failed login for user \"{}\": unknown user", username);
+                    Err(())
+                })
+                .or_else(|_| {
+                    // punish the failed login with a 1500ms delay before returning the error
+                    let when = Instant::now() + Duration::from_millis(1500);
```

Comment:
Security feature I assume. Cool!

Reply:
Yes, against brute force attacks. To make it solid, we should also limit the number of simultaneous connections from a single source IP though.

Reply:
nice